src/rules/matchers.ts

```diff
@@ -368,6 +374,68 @@ export class CookieMatcher extends Serializable implements RequestMatcher {
     }
 }
 
+export class CallbackMatcher extends Serializable implements RequestMatcher {
+    readonly type = "callback";
+  
+    constructor(
+        public callback: (request: OngoingRequest) => MaybePromise<boolean>
+        ) {
+        super();
+    }
+  
+    async matches(request: OngoingRequest) {
+        try {
+            return await this.callback(request);
+        } catch (error) {
+            return false;
+        }
+    }
+  
+    explain() {
+        return `matches using provided callback ${
+            this.callback.name ? ` (${this.callback.name})` : ""
+        }`;
+    }
+  
+    /**
+     * @internal
+     */
+    serialize(channel: ClientServerChannel): SerializedCallbackMatcherData {
+      channel.onRequest<OngoingRequest, boolean>(async (streamMsg) => {
+        const request = withDeserializedBodyReader(streamMsg as any);
+
+        const callbackResult = await this.callback.call(null, request as any);
```

Comment:
I think there's some type confusion here between CompletedRequest & OngoingRequest, which is why this `any` is required.

Right now, the types say that the callback must take an OngoingRequest, which is otherwise an internal type that's not usually exposed in the API. In `matches`, the raw OngoingRequest is given to the callback, but here for remote matching it's _partly_ converted (by withDeserializedBodyReader) to a CompletedRequest (i.e. the body becomes a CompletedBody).

Instead, I think this should probably just wait for the body & convert the OngoingRequest to a CompletedRequest before calling the callback with that. That shouldn't be too bad: there's a method called `waitForCompletedRequest` in request-utils which will wait and do all the proper conversion for you, and the existing serialization methods already handle CompletedRequests.

That would simplify things, and also has the advantage that the request body object in this method & `thenCallback` will be exactly the same - I feel like having different types there is quickly going to become confusing.

There is a minor performance impact to that approach (if you add a matcher like this then the resulting rule can't handle any requests until the entire body is received) but I think it's not a big deal, I suspect most custom callbacks will end up waiting for the body anyway, and the same thing unavoidably affects all the body matchers regardless.